core/services/relay/evm/chain_reader.go

```diff
@@ -23,132 +25,186 @@ type ChainReaderService interface {
-		}
-	}
+func (cr *chainReader) Close() error {
+	return cr.StopOnce("ChainReader", func() error {
+		return cr.contractBindings.ForEach(readBinding.Unregister)
+	})
+}
 
-	var chainReaderEventParams []string
-	for chainReaderEventParam := range chainReaderDefinition.Params {
-		chainReaderEventParams = append(chainReaderEventParams, chainReaderEventParam)
-	}
+func (cr *chainReader) Ready() error { return nil }
+func (cr *chainReader) HealthReport() map[string]error {
+	return map[string]error{cr.Name(): nil}
+}
 
-	if !areChainReaderArgumentsValid(abiEventIndexedInputs, chainReaderEventParams) {
-		var abiEventIndexedInputsNames []string
-		for _, abiEventIndexedInput := range abiEventIndexedInputs {
-			abiEventIndexedInputsNames = append(abiEventIndexedInputsNames, abiEventIndexedInput.Name)
-		}
-		return fmt.Errorf("params: [%s] don't match abi event indexed inputs: [%s]", strings.Join(chainReaderEventParams, ","), strings.Join(abiEventIndexedInputsNames, ","))
+func (cr *chainReader) CreateContractType(contractName, methodName string, forEncoding bool) (any, error) {
+	return cr.codec.CreateType(wrapItemType(contractName, methodName, forEncoding), forEncoding)
+}
+
+func wrapItemType(contractName, methodName string, isParams bool) string {
+	if isParams {
+		return fmt.Sprintf("params.%s.%s", contractName, methodName)
 	}
-	return nil
+	return fmt.Sprintf("return.%s.%s", contractName, methodName)
 }
 
-func validateMethods(abi abi.ABI, chainReaderDefinition types.ChainReaderDefinition) error {
+func (cr *chainReader) addMethod(
+	contractName,
+	methodName string,
+	abi abi.ABI,
+	chainReaderDefinition types.ChainReaderDefinition) error {
 	method, methodExists := abi.Methods[chainReaderDefinition.ChainSpecificName]
 	if !methodExists {
-		return fmt.Errorf("method: %q doesn't exist", chainReaderDefinition.ChainSpecificName)
+		return fmt.Errorf("%w: method %s doesn't exist", commontypes.ErrInvalidConfig, chainReaderDefinition.ChainSpecificName)
+	}
+
+	cr.contractBindings.AddReadBinding(contractName, methodName, &methodBinding{
+		contractName: contractName,
+		method:       methodName,
+		client:       cr.client,
+	})
+
+	if err := cr.addEncoderDef(contractName, methodName, method.Inputs, method.ID, chainReaderDefinition); err != nil {
+		return err
 	}
 
-	var methodNames []string
-	for methodName := range chainReaderDefinition.Params {
-		methodNames = append(methodNames, methodName)
+	return cr.addDecoderDef(contractName, methodName, method.Outputs, chainReaderDefinition)
+}
+
+func (cr *chainReader) addEvent(contractName, eventName string, a abi.ABI, chainReaderDefinition types.ChainReaderDefinition) error {
+	event, eventExists := a.Events[chainReaderDefinition.ChainSpecificName]
+	if !eventExists {
+		return fmt.Errorf("%w: method %s doesn't exist", commontypes.ErrInvalidConfig, chainReaderDefinition.ChainSpecificName)
+	}
+	cr.contractBindings.AddReadBinding(contractName, eventName, &eventBinding{
+		lp:   cr.lp,
```

Comment:
shouldn't we also set contract name and event name here?

Reply:
It doesn't need it, it doesn't make a call on-chain.

I'm working on another PR that will allow indexed fields to work and be searched, it'll need it then to make life easier.

Reply:
The event binding didn't need it.